### PR TITLE
8337225: Demote maxStack and maxLocals from CodeModel to CodeAttribute

### PR DESCRIPTION
**Pull request by liach**
As discussed in offline meeting, the max stack and locals information are part of the code attribute and not meaningful for buffered code elements. Computation would be costly and these see no real usage during transformations. Thus, the proposed solution is to move these APIs to be CodeAttribute specific, as this is already how all these APIs' users are using.

Also removed useless `Writable` on buffered models, and fixed `BufferedMethodBuilder::code` implementation.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8337238](https://bugs.openjdk.org/browse/JDK-8337238) to be approved

### Issues
 * [JDK-8337225](https://bugs.openjdk.org/browse/JDK-8337225): Demote maxStack and maxLocals from CodeModel to CodeAttribute (**Bug** - P4)
 * [JDK-8337238](https://bugs.openjdk.org/browse/JDK-8337238): Demote maxStack and maxLocals from CodeModel to CodeAttribute (**CSR**)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20338/head:pull/20338` \
`$ git checkout pull/20338`

Update a local copy of the PR: \
`$ git checkout pull/20338` \
`$ git pull https://git.openjdk.org/jdk.git pull/20338/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20338`

View PR using the GUI difftool: \
`$ git pr show -t 20338`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20338.diff">https://git.openjdk.org/jdk/pull/20338.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20338#issuecomment-2251524615)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/BufferedCodeBuilder.java [hunk 1]
```diff
@@ -60,11 +57,10 @@ public BufferedCodeBuilder(MethodInfo methodInfo,
         this.context = context;
         this.startLabel = new LabelImpl(this, -1);
         this.endLabel = new LabelImpl(this, -1);
-        this.original = original;
         this.methodInfo = methodInfo;
         this.maxLocals = Util.maxLocals(methodInfo.methodFlags(), methodInfo.methodTypeSymbol());
-        if (original != null)
-            this.maxLocals = Math.max(this.maxLocals, original.maxLocals());
```
Comment (asotona) [inline src/java.base/share/classes/jdk/internal/classfile/impl/BufferedCodeBuilder.java:67]: `original::maxLocals`set the counter for `CodeBuilder::allocateLocal`
By restricting calculation of maxLocals to "origin instanceof CodeAttribute" may cause invalid locals allocations for chained code builders. The number might not be exposed in the API, however we need to know it internally.

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/BufferedCodeBuilder.java:67]: I think we should just do `Util.maxLocals` without this `origin` check. Your said problem can happen too if a DirectMethodBuilder transforms a BufferedCodeBuilder.Model.

Note that any code builder can receive a code model half way with `cob.transform(anyModel, CodeTransform.ACCEPT_ALL)`, which the buffered code builder initialization also uses.
I think we should just ask users to use `CodeLocalsShifter` in that case, especially if the transform is inserting new locals. We cannot buffer existing chained builders for performance reasons.

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/BufferedCodeBuilder.java:67]: Updated this to internal API that accesses the existing local variable management on BufferedCodeBuilder. We can discuss this consistency problem later.

File: src/java.base/share/classes/jdk/internal/classfile/impl/TerminalCodeBuilder.java [hunk 2]
```diff
@@ -25,8 +25,24 @@
...
         permits DirectCodeBuilder, BufferedCodeBuilder {
     int curTopLocal();
+
+    static int setupTopLocal(MethodInfo methodInfo, CodeModel original) {
+        int paramSlots = Util.maxLocals(methodInfo.methodFlags(), methodInfo.methodTypeSymbol());
+        if (original == null) {
+            return paramSlots;
+        }
+        if (original instanceof CodeAttribute attr) {
+            return Math.max(paramSlots, attr.codeLength());
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/TerminalCodeBuilder.java:41]:
```suggestion
            return Math.max(paramSlots, attr.maxLocals());
```

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/TerminalCodeBuilder.java:41]: Good point, will add a test too to verify.

File: src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java [hunk 3]
```diff
@@ -312,8 +310,9 @@ public void writeBody(BufWriterImpl b) {
 
             private void writeCounters(boolean codeMatch, BufWriterImpl buf) {
                 if (codeMatch) {
-                    buf.writeU2(original.maxStack());
-                    buf.writeU2(original.maxLocals());
+                    var originalAttribute = (CodeImpl) original;
```
Comment (asotona) [inline src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java:313]: Are you absolutely sure a `BufferedCodeBuilder.Model` cannot appear here?

**Review by asotona — APPROVED** [review]
Looks good to me.

File: test/jdk/jdk/classfile/BuilderBlockTest.java [hunk 4]
```diff
@@ -305,4 +317,81 @@ void testAllocateLocalIfThen() {
...
+            assertTrue(foundItem);
+        }
+    });
+
+    // Test updating local variable slot management from
+    // source code models in transformingCode;
+    // CodeBuilder.transform(CodeModel, CodeTransform) is
+    // not managed for now
+    @Test
+    void testAllocateLocalTransformingCodeAttribute() throws IOException  {
```
Comment (turbanoff) [inline test/jdk/jdk/classfile/BuilderBlockTest.java:353]:
```suggestion
    void testAllocateLocalTransformingCodeAttribute() throws IOException {
```

**Review by asotona — APPROVED** [review]

**Comment by liach** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**